### PR TITLE
Save git flow branches in tasks, so IDEA can automatically change bra…

### PR DESCRIPTION
**Pull request by YanickFrehner**
…nches when switching between tasks.

### CODE REVIEW

**Comment by OpherV** [comment]
Thanks @YanickFrehner! I will review

**Comment by YanickFrehner** [comment]
Are there any news regarding this Pull request?

**Comment by OpherV** [comment]
Yes! I'm away giving a talk in JSConf Iceland but I'll be back next week and merge it then 😃

**Comment by yanickch** [comment]
Ok nice, what is the talk about you are giving?

**Comment by YanickFrehner** [comment]
Ups, wrong account :)

**Comment by OpherV** [comment]
Amazing, weird and things things which aren't websites that you can do with JS :)
https://2018.jsconf.is/speakers/opher-vishnia/

You can watch the conf live if you'd like
https://2018.jsconf.is/live/

File: src/gitflow/actions/GitflowAction.java [hunk 1]
```diff
@@ -16,6 +16,7 @@
 import gitflow.GitflowConfigUtil;
 import gitflow.ui.NotifyUtil;
 
+import javax.annotation.Nullable;
```
Comment (OpherV) [inline src/gitflow/actions/GitflowAction.java]: Please use `import org.jetbrains.annotations.Nullable;`

File: src/gitflow/actions/StartFeatureAction.java [hunk 2]
```diff
@@ -15,6 +13,8 @@
 import gitflow.ui.GitflowStartFeatureDialog;
 import gitflow.ui.NotifyUtil;
 
+import javax.annotation.Nullable;
```
Comment (OpherV) [inline src/gitflow/actions/StartFeatureAction.java]: Please use `import org.jetbrains.annotations.Nullable`;

File: src/gitflow/actions/StartHotfixAction.java [hunk 3]
```diff
@@ -13,6 +13,8 @@
 import gitflow.ui.GitflowStartHotfixDialog;
 import gitflow.ui.NotifyUtil;
 
+import javax.annotation.Nullable;
```
Comment (OpherV) [inline src/gitflow/actions/StartHotfixAction.java]: Please use `import org.jetbrains.annotations.Nullable`;

File: src/gitflow/ui/GitflowOpenTaskPanel.java [hunk 4]
```diff
@@ -100,19 +106,34 @@ public void commit() {
...
-            action =  new StartHotfixAction(myRepo);
-            action.runAction(myProject, selectedHotfixBaseBranch.getBranchName(), hotfixName.getText());
-            gitflowState.setTaskBranch(currentTask, GitflowConfigUtil.getHotfixPrefix(myProject, myRepo) + hotfixName.getText());
+            final String branchName = GitflowConfigUtil.getHotfixPrefix(myProject, myRepo) + hotfixName.getText();
+            runAction(new StartHotfixAction(myRepo), selectedHotfixBaseBranch.getBranchName(), branchName);
+        }
+    }
+
+    private void runAction(GitflowAction action, String baseBranchName, final String branchName) {
+        final TaskInfo[] current = myVcsTaskHandler.getCurrentTasks();
```
Comment (OpherV) [inline src/gitflow/ui/GitflowOpenTaskPanel.java]: Please add some comments to document the logic in this function

File: src/gitflow/ui/GitflowOpenTaskPanel.java [hunk 5]
```diff
@@ -100,19 +106,34 @@ public void commit() {
...
         else if (startHotfixRadioButton.isSelected()) {
-            action =  new StartHotfixAction(myRepo);
-            action.runAction(myProject, selectedHotfixBaseBranch.getBranchName(), hotfixName.getText());
-            gitflowState.setTaskBranch(currentTask, GitflowConfigUtil.getHotfixPrefix(myProject, myRepo) + hotfixName.getText());
+            final String branchName = GitflowConfigUtil.getHotfixPrefix(myProject, myRepo) + hotfixName.getText();
+            runAction(new StartHotfixAction(myRepo), selectedHotfixBaseBranch.getBranchName(), branchName);
+        }
+    }
+
+    private void runAction(GitflowAction action, String baseBranchName, final String branchName) {
```
Comment (OpherV) [inline src/gitflow/ui/GitflowOpenTaskPanel.java]: I would call this wrapper something like `attachTaskAndRunAction`. Verbose, I know, but I want to differentiate this `runAction` from the `GitflowAction`'s  `runAction` so it's not as confusing

File: src/gitflow/ui/GitflowOpenTaskPanel.java [hunk 6]
```diff
@@ -100,19 +106,34 @@ public void commit() {
...
+        final TaskInfo[] current = myVcsTaskHandler.getCurrentTasks();
+
+        if (myPreviousTask != null && myPreviousTask.getBranches(false).isEmpty()) {
+            TaskManagerImpl.addBranches(myPreviousTask, current, false);
```
Comment (OpherV) [inline src/gitflow/ui/GitflowOpenTaskPanel.java]: What if the previous task wasn't associated with this branch? Maybe it wasn't associated with any branch in particular... I think it's dangerous to assume that if a current task doesn't have an associated branch then you auto-attach it to the current branch. 

The very least we could have a confirmation popup that tells the user "You are about to switch to a different task, which would change the current branch. Would you like to associate the current task with the current branch before switching?"

**Pull request closed without merging**